java/source/flight.rst

```diff
@@ -0,0 +1,1050 @@
+For more detail about Flight please take a look at `Arrow Flight RPC`_.
+
+.. contents::
+
+Simple Key-Value Storage Service with Arrow Flight
+==================================================
+
+We'll implement a service that provides a key-value store for data, using Flight to handle uploads/requests
+and data in memory to store the actual data.
+
+Flight Client and Server
+************************
+
+.. code-block:: java
+
+    import org.apache.arrow.flight.Action;
+    import org.apache.arrow.flight.AsyncPutListener;
+    import org.apache.arrow.flight.CallStatus;
+    import org.apache.arrow.flight.Criteria;
+    import org.apache.arrow.flight.FlightClient;
+    import org.apache.arrow.flight.FlightDescriptor;
+    import org.apache.arrow.flight.FlightEndpoint;
+    import org.apache.arrow.flight.FlightInfo;
+    import org.apache.arrow.flight.FlightServer;
+    import org.apache.arrow.flight.FlightStream;
+    import org.apache.arrow.flight.Location;
+    import org.apache.arrow.flight.NoOpFlightProducer;
+    import org.apache.arrow.flight.PutResult;
+    import org.apache.arrow.flight.Result;
+    import org.apache.arrow.flight.Ticket;
+    import org.apache.arrow.memory.RootAllocator;
+    import org.apache.arrow.vector.VarCharVector;
+    import org.apache.arrow.vector.VectorLoader;
+    import org.apache.arrow.vector.VectorSchemaRoot;
+    import org.apache.arrow.vector.VectorUnloader;
+    import org.apache.arrow.vector.ipc.message.ArrowRecordBatch;
+    import org.apache.arrow.vector.types.pojo.ArrowType;
+    import org.apache.arrow.vector.types.pojo.Field;
+    import org.apache.arrow.vector.types.pojo.FieldType;
+    import org.apache.arrow.vector.types.pojo.Schema;
+
+    import java.io.IOException;
+    import java.nio.charset.StandardCharsets;
+    import java.util.ArrayList;
+    import java.util.Arrays;
+    import java.util.Collections;
+    import java.util.Iterator;
+    import java.util.List;
+    import java.util.Map;
+    import java.util.concurrent.ConcurrentHashMap;
+
+    class DataInMemory {
+        private List<ArrowRecordBatch> listArrowRecordBatch;
+        private Schema schema;
+        private long rows;
+        public DataInMemory(List<ArrowRecordBatch> listArrowRecordBatch, Schema schema, long rows) {
+            this.listArrowRecordBatch = listArrowRecordBatch;
+            this.schema = schema;
+            this.rows = rows;
+        }
+        public List<ArrowRecordBatch> getListArrowRecordBatch() {
+            return listArrowRecordBatch;
+        }
+        public Schema getSchema() {
+            return schema;
+        }
+        public long getRows() {
+            return rows;
+        }
+    }
+
+    // Server
+    Location location = Location.forGrpcInsecure("0.0.0.0", 33333);
+    Map<FlightDescriptor, DataInMemory> dataInMemory = new ConcurrentHashMap<>();
+    try (RootAllocator allocator = new RootAllocator(Long.MAX_VALUE)){
+        class CookbookProducer extends NoOpFlightProducer {
```

Comment:
Can't we move this up alongside DataInMemory? (We can add a constructor to give it any variables needed.)

Reply:
Yes we can, but another class is needed to create, because current design CookbookProducer update DataInMemory data. Could be possible maintain as it is now?

Reply:
Why? `dataInMemory` should be a property of this class. We shouldn't be using globals from the server.